evalml/automl/engine/engine_base.py

```diff
@@ -0,0 +1,142 @@
+import sys
+import time
+import traceback
+from abc import ABC, abstractmethod
+from collections import OrderedDict
+
+import numpy as np
+import pandas as pd
+
+from evalml.automl.utils import tune_binary_threshold
+from evalml.exceptions import PipelineScoreError
+from evalml.model_family import ModelFamily
+from evalml.preprocessing import split_data
+from evalml.problem_types import ProblemTypes, is_binary, is_multiclass
+from evalml.utils.logger import get_logger
+from evalml.utils.woodwork_utils import _convert_woodwork_types_wrapper
+
+logger = get_logger(__file__)
+
+
+class EngineBase(ABC):
+    """Base class for the engine API which handles the fitting and evaluation of pipelines during AutoML."""
+
+    def __init__(self, automl=None, should_continue_callback=None, pre_evaluation_callback=None, post_evaluation_callback=None):
+        """Base class for the engine API which handles the fitting and evaluation of pipelines during AutoML.
+
+        Arguments:
+            automl (AutoMLSearch): a reference to the AutoML search. Used to access configuration and by the error callback.
+            should_continue_callback (function): returns true if another pipeline from the list should be evaluated, false otherwise.
+            pre_evaluation_callback (function): optional callback invoked before pipeline evaluation.
+            post_evaluation_callback (function): optional callback invoked after pipeline evaluation, with args pipeline and evaluation results. Expected to return a list of pipeline IDs corresponding to each pipeline evaluation.
+        """
+        self.automl = automl
```

Comment:
This code still asks for a direct reference to `AutoMLSearch`, even though the callbacks mostly don't depend on that direct access. Two reasons for that:
* The `AutoMLSearch` error callback currently expects a reference to the `AutoMLSearch` so it can optionally add the error to `self._results`. We should look into changing this but I didn't wanna do it in this PR.
* We don't have a config object yet, so it was the easiest way to pass things like `problem_type` and `objective` to the static method `train_and_score_pipeline` below